src/d3a/models/area/markets.py

```diff
@@ -15,84 +15,158 @@
+
+    def _move_markets_to_past(self, current_time: DateTime) -> None:
+        """Move slot markets to self.past_markets."""
+        # the conversion to list is needed in order to omit 'mutated during iteration' error
+        for time_slot in list(self.markets.keys()):
+            if time_slot < current_time:
+                market = self.markets.pop(time_slot)
                 market.readonly = True
-                self._delete_past_markets(past_markets)
-                past_markets[timeframe] = market
-                self.log.trace("Moving {t:%H:%M} {m} to past"
-                               .format(t=timeframe, m=past_markets[timeframe].name))
-
-    def _delete_past_markets(self, past_markets):
-        if not d3a.constants.RETAIN_PAST_MARKET_STRATEGIES_STATE:
-            delete_markets = [pm for pm in past_markets if
-                              pm not in self.markets.values()]
-            for pm in delete_markets:
-                if ConstSettings.GeneralSettings.EVENT_DISPATCHING_VIA_REDIS:
-                    past_markets[pm].redis_api.stop()
-                del past_markets[pm].offers
-                del past_markets[pm].trades
-                del past_markets[pm].offer_history
-                del past_markets[pm].notification_listeners
-                del past_markets[pm].bids
-                del past_markets[pm].bid_history
-                del past_markets[pm].traded_energy
-                del past_markets[pm]
+                self.past_markets[time_slot] = market
+                self.log.debug(
+                    f"Moving {self.past_markets[time_slot]} to past.")
+
+    def _move_balancing_markets_to_past_balancing(self, current_time: DateTime) -> None:
+        """Move balancing markets to self.past_balancing_markets."""
+        # the conversion to list is needed in order to omit 'mutated during iteration' error
+        for time_slot in list(self.balancing_markets.keys()):
+            if time_slot < current_time:
+                market = self.balancing_markets.pop(time_slot)
+                market.readonly = True
+                self.past_balancing_markets[time_slot] = market
+                self.log.debug(
+                    f"Moving {self.past_balancing_markets[time_slot]} to past.")
+
+    def _move_settlement_markets_to_past_settlement(self, current_time: DateTime) -> None:
+        """Move settlement markets to self.past_settlement_markets."""
+        move_settlement_market_slots = [
+            time_slot for time_slot in self.settlement_markets.keys()
+            if self._is_it_time_to_delete_past_market(current_time, time_slot)]
+        for time_slot in move_settlement_market_slots:
+            self.past_settlement_markets[time_slot] = self.settlement_markets.pop(time_slot)
+            self.log.debug(
+                f"Moving {self.past_settlement_markets[time_slot]} to past.")
+
+    def _delete_past_markets(self, current_time: DateTime) -> None:
```

Comment:
So we have the following similar methods:

1. _delete_past_markets
2. _delete_past_balancing_markets
3. _delete_past_settlement_markets

The first two have the exact same implementation (but each one of them refer to a different set of markets), whereas the third one checks the `_is_it_time_to_delete_past_settlement_market` instead of the `_is_it_time_to_delete_past_market`

Can we refactor these three to have only one method responsible for deleting the past markets?
we may need to pass the relevant markets to this new method (either markets, balancing markets or settlement markets).
+ for the third case we may need to update the `_is_it_time_to_delete_past_market` to cover the case of the settlement markets.

Otherwise the code is linearly scaling up for each new market type that we will add, let me know what you think about that.

Reply:
I am aware of the code duplication here and wanted to leave a comment about this, but forgot.
I decided to keep this because of [Pierpaolos earlier comment](https://github.com/gridsingularity/d3a/pull/1213#discussion_r696477792).
I also kept this way to have a better readability and remembered the workshop where we quoted Sandi Metz
`Duplication is far cheaper than the wrong abstraction`. This part of the code will hopefully not be touched anytime in the future (it wasn't for a long time until now). So IMO, we can live with the duplication for better readability and debugging. Happy to discuss this and hear other opinions.

Reply:
Of course duplication is cheaper than wrong abstraction, but accurate abstraction is far better than duplication as well :P so if we can agree on the abstraction it would be great, this way you can also reduce the number of code lines here + the unittests functions

Reply:
I would personally refactor it. However, if you think that refactoring these methods is not worth it that's fine. Still, I think it would be better to have them all in one place (e.g. all in `util.py`). That's because they're doing the same thing, and if we decide to refactor them in the future it will be useful to have them all in one place at least.

Reply:
I would agree to refactor them, I would not agree to move these to util.py. This inherently belongs to the market class, and taking it out of the way because it looks boilerplateish is not the right way IMO